src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+        CRestrictedReuseAreaStack(const CRestrictedReuseAreaStack&) = delete;
+        CRestrictedReuseAreaStack& operator = (const CRestrictedReuseAreaStack&) = delete;
+        ~CRestrictedReuseAreaStack() = default;
+
+        void PushToBack(RestrictedReuseArea area);
+        RestrictedReuseArea PopFromBack();
+        RestrictedReuseArea& PeekBack();
+        // TODO: Remove and use size directly?
+        int Count();
+    };
+
+    class CQubitManagerRestrictedReuse : public IQubitManager
+    {
+        // Restricted reuse area control, allocation and release
+
+        public: void StartRestrictedReuseArea();
+        public: void NextRestrictedReuseSegment();
+        public: void EndRestrictedReuseArea();
+        private: long AllocateQubit(); // Computation complexity is O(number of nested restricted reuse areas).
+        private: void ReleaseQubit(long id);
+
+        // Configuration Properties
+
+        // TODO: Make sure these are read only
+        public: bool MayExtendCapacity;
+        public: bool EncourageReuse;
+
+        // Constructors/destructors, storage and reallocation
+
+        private: int* sharedQubitStatusArray; // Tracks allocation state of all qubits. Stores lists of free qubits.
+        private: int qubitCapacity = 0; // qubitCapacity is always equal to the array size.
+        private: CRestrictedReuseAreaStack freeQubitsInAreas; // Fresh Free Qubits are located in freeQubitsInAreas[0].FreeQubitsReuseAllowed
+        
+        public: CQubitManagerRestrictedReuse(
+            int initialQubitCapacity,
+            bool mayExtendCapacity = false,
+            bool encourageReuse = true);
+
+        // No complex scenarios for now. Don't need to support copying/moving.
+        public: CQubitManagerRestrictedReuse(const CQubitManagerRestrictedReuse&) = delete;
+        public: CQubitManagerRestrictedReuse& operator = (const CQubitManagerRestrictedReuse&) = delete;
+        public: ~CQubitManagerRestrictedReuse();
+        private: void EnsureCapacity(int requestedCapacity);
+
+        // Disable, disabled qubit count and checks
+
+        public: int DisabledQubitsCount = 0;
+        public: bool IsDisabled(int id);
+        public: bool IsDisabled(Qubit qubit);
+        // Disables a given qubit.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        public: void Disable(Qubit qubit);
+        // Disables a set of given qubits.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        public: void Disable(Qubit* qubitsToDisable, int qubitCount);
+
+        // Allocate, allocated qubit count and checks
+
+        public: int AllocatedQubitsCount = 0;
+        public: bool IsExplicitlyAllocated(long id);
+        public: void ChangeStatusToAllocated(long id);
+        // Allocates a qubit. Extend capacity if necessary and possible.
+        // Fails if the qubit cannot be allocated. 
+        public: Qubit Allocate();
```

Comment:
`Qubit Allocate()` and `void Release(Qubit qubit)` are `CQubitManagerRestrictedReuse`'s overrides of the virtual functions inherited from `IQubitManager`:
```c++
virtual Qubit Allocate() = 0;        
virtual void Release(Qubit qubit) = 0;
```
Consider using the keyword `override` for the overrides. See [Item 12: Declare overriding functions `override`](https://moodle.ufsc.br/pluginfile.php/2377667/mod_resource/content/0/Effective_Modern_C__.pdf)

Reply:
IQubitManager was an potential integration point. I don't think we need it at the moment as it just gets in the way. When we figure out our extensibility story we will extract the interface. I removed IQubitManager for now.